### PR TITLE
Approximate hessian matrix using central differences

### PR DESCRIPTION
**Pull request by skavulya**
This is an initial implementation of the approximate hessian matrix using the central difference equation. It is the same implementation used by optimHess() in the R stats package. I used the example in R for the initial  test case. I am new to Breeze so I might have used the wrong data structures and conventions. Please let me know what changes I need to make.

### CODE REVIEW

File: math/src/main/scala/breeze/optimize/SecondOrderFunction.scala [hunk 1]
```diff
@@ -68,38 +68,85 @@ object SecondOrderFunction {
...
+    // second order differential using central differences
+    for (i <- 0 until n) {
+      x(i) = x(i) + eps
+      val df1 = df.gradientAt(x)
+
+      x(i) = x(i) - 2 * eps
+      val df2 = df.gradientAt(x)
+
+      val grad = (df1 - df2)/(2*eps)
+      for (j <- 0 until n) {
```
Comment (dlwh) [inline math/src/main/scala/breeze/optimize/SecondOrderFunction.scala]: this loop can just be H(i, ::) := grad

Comment (skavulya) [inline math/src/main/scala/breeze/optimize/SecondOrderFunction.scala]: Thanks. I made the change, and updated the code to use a copy of the input vector. I was thinking of refactoring my code, and creating CentralDiffFunction class which extends SecondOrderFunction. The hessian calculation would be done in calculate2(). However, I am not sure how to represent the matrix in Breeze. Would it be an IndexedSeq[T]? If so, is there a way I can convert the IndexedSeq to a matrix?

**Comment by dlwh** [comment]
thanks!

File: math/src/main/scala/breeze/optimize/SecondOrderFunction.scala (hunk 1 above)
Comment (dlwh) [inline math/src/main/scala/breeze/optimize/SecondOrderFunction.scala]: for now I'd just force T to be DenseVector

**Comment by skavulya** [comment]
Thanks for all your help!

**Pull request merged**